include/deal.II/base/mpi_compute_index_owner_internal.h

```diff
@@ -0,0 +1,799 @@
+          {
+            Assert(recv_indices[other_rank].size() == recv_buffer.size(),
+                   ExcMessage("Sizes do not match!"));
+
+            for (unsigned int j = 0; j < recv_indices[other_rank].size(); j++)
+              owning_ranks[recv_indices[other_rank][j]] = recv_buffer[j];
+          }
+
+          /**
+           * Resolve the origin of the requests by sending the information
+           * accumulated in terms of the dictionary owners during the run of
+           * the consensus algorithm back to the owner in the original
+           * IndexSet. This requires some point-to-point communication.
+           *
+           * @return Map of processors and associated ranges of indices that
+           *         are requested from the current rank. The index ranges are
+           *         expressed in the local index space associated to the
+           *         locally_owned indices of this rank.
+           */
+          std::map<unsigned int, std::vector<std::pair<unsigned, unsigned int>>>
+          get_requesters()
+          {
+            Assert(track_index_requests,
+                   ExcMessage("Must enable index range tracking in"
+                              "constructor of ConsensusAlgorithmProcess"));
+
+            std::map<unsigned int,
+                     std::vector<std::pair<unsigned, unsigned int>>>
+              requested_indices;
+
+#ifdef DEAL_II_WITH_MPI
+
+            // reserve enough slots for the requests ahead; depending on
+            // whether the owning rank is one of the requesters or not, we
+            // might have one less requests to execute, so fill the requests
+            // on demand.
+            std::vector<MPI_Request> send_requests;
+            send_requests.reserve(requesters.size());
+
+            // We use an integer vector for the data exchange. Since we send
+            // data associated to intervals with different requesters, we will
+            // need to send (a) the MPI rank of the requester, (b) the number
+            // of intervals directed to this requester, and (c) a list of
+            // intervals, i.e., two integers per interval. The number of items
+            // sent in total can be deduced both via the MPI status message at
+            // the receiver site as well as be counting the buckets from
+            // different requesters. First we count how much data to reserve
+            // to make sure no memory-reallocation is needed while filling in
+            // the data and communicating, as we need to keep the buffers
+            // alive while MPI is active.
+            std::vector<unsigned int> send_data;
+            {
+              unsigned int n_integers = 0;
+              for (unsigned int i = 0; i < requesters.size(); ++i)
+                if (dict.actually_owning_rank_list[i] != my_rank)
+                  for (const auto &j : requesters[i])
+                    n_integers += 2 + 2 * j.second.size();
```

Comment:
I find packing data manually into a ``vector<unsigned int>`` hard to understand/debug/fix. I understand that you don't want the overhead of serializing a data structure. Would it be difficult to have an internal class that stores data contiguously internally?

Reply:
The question is whether the additional class is worth the effort, since it's a single place. A step into this direction would be to convert the `std::vector<unsinged int>` into an `std::vector<std::pair<unsigned int,unsigned int>>`. Then we still have the ambiguity that the very first entry interprets the pair as the number of "queued pairs for this process" plus the processor rank, but at least the other items are ranges as intended. At least that's what we do in the respective classes of `DoFHandlerPolicy`: We misuse the first integer in some global dof numbers to indicate how many dof numbers follow for the current element (to support hp). That's no excuse for the sin of course, but I would rather live with this "packing" problem given the local scope than an additional class that hides the packing.

Reply:
> Then we still have the ambiguity that the very first entry interprets the pair as the number of "queued pairs for this process" plus the processor rank, but at least the other items are ranges as intended. 
That doesn't sound better to me. :-)

I wish there was an easy way to have several std::vectors place their memory allocations contiguously after each other.

Reply:
You can always write a custom allocator. :wink:

Reply:
If they're the same size, you can always use `Table<2,T>`.

Reply:
No, unfortunately not. Let us stick with the current solution, given that I'm unhappy to move the mechanism to stick things together, used once here, to a separate class.